tests/testthat/error/test-conditions.txt

```diff
@@ -101,15 +101,15 @@ lossy cast from character to factor mentions loss of generality
 ===============================================================
 
 > vec_cast("a", factor("b"))
-Error: Can't convert from <character> to <factor<ddf10>> due to loss of generality.
+Error: Can't convert from <character> to <factor<9b7e3>> due to loss of generality.
```

Comment:
Is this now being OS/R version dependent going to cause problems?

Reply:
It should not be OS dependent, so that should be okay.

For somewhat complex types it may be R version dependent (I noticed it was with data frames), but simple R objects seem to give the same hash across R versions. The tests pass here, so it seems like factors are okay?

I'm not sure what is different about data frames, but R core must have changed something about the way they are structured somewhere along the way, which changed the hash - possibly they changed something about the way internal row names work?

Reply:
Oh I figured out what changed with data frames. Some time between R 3.3 and now, the order of the attributes changed. `class` now comes before `row.names`, and that changed the internal hash. Otherwise, barring ALTREP differences, the hashes are _generally_ also stable across R versions, but we can't promise that because of things like this.

R 3.3.3:

```r
> digest::digest(data.frame(x=1))
[1] "e74d491629d919e23692a94cd9e5b935"

> digest::digest(structure(list(x = 1), row.names = c(NA, -1L), class = "data.frame"))
[1] "e74d491629d919e23692a94cd9e5b935"

> digest::digest(structure(list(x = 1), class = "data.frame", row.names = c(NA, -1L)))
[1] "65a37f3d72f58de9eb80cd1674896a60"
```

R 4.0.3

```r
> digest::digest(data.frame(x=1))
[1] "65a37f3d72f58de9eb80cd1674896a60"

> digest::digest(structure(list(x = 1), row.names = c(NA, -1L), class = "data.frame"))
[1] "e74d491629d919e23692a94cd9e5b935"

> digest::digest(structure(list(x = 1), class = "data.frame", row.names = c(NA, -1L)))
[1] "65a37f3d72f58de9eb80cd1674896a60"
```